### PR TITLE
Change autoupdate proto messages

### PR DESCRIPTION
**Pull request by hugoShaka**
This commits does 3 changes:
- reflect the maintenance duration on the rollout in a new spec field
- add the rollout start time field in its status
- change wait_days into wait_hours

Part of: https://github.com/gravitational/teleport/pull/47126

Goal (internal): https://github.com/gravitational/cloud/issues/10289

Context:
- I need to add the maintenance duration in the rollout to support customizing maintenance duration (current implem just does 1h windows) and forgot this in the RFD
- We must track the rollout start time to add logic that avoids awkward situations when a group start immediately when we change the version.
- Wait_days behaviour is ambiguous in the RFD (do we count 24 hours since the group start, or check midnight UTC?, midnight local tz?). Switchig to hours solves the issue

### CODE REVIEW

**Review by sclevine — APPROVED** [review]

**Review by vapopov — APPROVED** [review]

File: api/proto/teleport/autoupdate/v1/autoupdate.proto [hunk 1]
```diff
@@ -71,14 +71,17 @@ message AgentAutoUpdateSchedules {
...
+
   // name of the group
   string name = 1;
   // days when the update can run. Supported values are "Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun" and "*"
   repeated string days = 2;
   // start_hour to initiate update
   int32 start_hour = 3;
-  // wait_days after last group succeeds before this group can run. This can only be used when the strategy is "halt-on-failure".
-  int64 wait_days = 4;
+  // wait_hours after last group succeeds before this group can run. This can only be used when the strategy is "halt-on-failure".
```
Comment (codingllama) [inline api/proto/teleport/autoupdate/v1/autoupdate.proto:83]: Are zero or negative hours accepted? Is that worth mentioning?

Have you considered `google.protobuf.Duration wait_duration`?

File: api/proto/teleport/autoupdate/v1/autoupdate.proto [hunk 2]
```diff
@@ -159,16 +162,31 @@ message AutoUpdateAgentRolloutSpec {
...
+  // maintenance window.
+  google.protobuf.Duration maintenance_window_duration = 6;
 }
 
 // AutoUpdateAgentRolloutStatus tracks the current agent rollout status.
 message AutoUpdateAgentRolloutStatus {
   repeated AutoUpdateAgentRolloutStatusGroup groups = 1;
   AutoUpdateAgentRolloutState state = 2;
+  // The start time is set when the rollout is created or reset. Usually this is caused by a version change.
+  // This timestamps allows the controller to detect that the rollout just changed.
```
Comment (codingllama) [inline api/proto/teleport/autoupdate/v1/autoupdate.proto:176]:
```suggestion
  // The timestamp allows the controller to detect that the rollout just changed.
```

**Review by zmb3 — APPROVED** [review]

File: api/proto/teleport/autoupdate/v1/autoupdate.proto [hunk 3]
```diff
@@ -183,8 +201,8 @@ message AutoUpdateAgentRolloutStatusGroup {
   repeated string config_days = 6;
   // config_start_hour to initiate update
   int32 config_start_hour = 7;
-  // config_wait_days after last group succeeds before this group can run. This can only be used when the strategy is "halt-on-failure".
-  int64 config_wait_days = 8;
+  // config_wait_hours after last group succeeds before this group can run. This can only be used when the strategy is "halt-on-failure".
+  int64 config_wait_hours = 9;
```
Comment (codingllama) [inline api/proto/teleport/autoupdate/v1/autoupdate.proto:205]: Same questions here regarding zero, negative, and google.protobuf.Duration.

File: api/proto/teleport/autoupdate/v1/autoupdate.proto [hunk 4]
```diff
@@ -71,14 +71,17 @@ message AgentAutoUpdateSchedules {
 
 // AgentAutoUpdateGroup specifies the update schedule for a group of agents.
 message AgentAutoUpdateGroup {
+  reserved 4;
+  reserved "wait_days";
```
Comment (codingllama) [inline api/proto/teleport/autoupdate/v1/autoupdate.proto:75]: +1 for the clarity of changing from days to an unambiguous time unit. Nicely done.

File: api/proto/teleport/autoupdate/v1/autoupdate.proto (hunk 1 above)
Comment (hugoShaka) [reply api/proto/teleport/autoupdate/v1/autoupdate.proto:83]: > Are zero or negative hours accepted? Is that worth mentioning?

No, I'll reject them during validation. Yes I'll add it.

> Have you considered google.protobuf.Duration wait_duration?

Yes but we do have a max resolution of an hour with `start_hour` so it makes more sense to keep the same unit.

File: api/proto/teleport/autoupdate/v1/autoupdate.proto [hunk 5]
```diff
@@ -159,16 +162,31 @@ message AutoUpdateAgentRolloutSpec {
   //   availability. A group finishes its update once most of its agents are running the correct version. Agents that
   //   missed the group update will try to catch back as soon as possible.
   string strategy = 5;
+  // maintenance_window_duration is the maintenance window duration. This can only be set if `strategy` is "time-based".
+  // Once the window is over, the group transitions to the done state. Existing agents won't be updated until the next
+  // maintenance window.
+  google.protobuf.Duration maintenance_window_duration = 6;
```
[the remark this answers is not in the document]
Comment (hugoShaka) [reply api/proto/teleport/autoupdate/v1/autoupdate.proto:168]: No, this is a hard constraint.
>   // Once the window is over, the group transitions to the done state. 

This is the concept behind the time-base strategy. For users who want to trade maintenance window against ordered and dependent updates, we have the "halt-on-error" strategy. The difference is documented there:

https://github.com/gravitational/teleport/blob/363f83abc82ebc729614bebfdcf5367310eced16/api/proto/teleport/autoupdate/v1/autoupdate.proto#L151-L160

[the remark this answers is not in the document]
Comment (codingllama) [reply api/proto/teleport/autoupdate/v1/autoupdate.proto:168]: Thanks!

File: api/proto/teleport/autoupdate/v1/autoupdate.proto [hunk 6]
```diff
@@ -199,7 +219,7 @@ enum AutoUpdateAgentGroupState {
   AUTO_UPDATE_AGENT_GROUP_STATE_ACTIVE = 2;
   // AUTO_UPDATE_AGENT_GROUP_STATE_DONE represents that the group has been updated. New agents should run v2.
   AUTO_UPDATE_AGENT_GROUP_STATE_DONE = 3;
-  // AUTO_UPDATE_AGENT_GROUP_STATE_ROLLEDBACK represents that the group has been rolleback.
+  // AUTO_UPDATE_AGENT_GROUP_STATE_ROLLEDBACK represents that the group has been rolledback.
```
Comment (codingllama) [inline api/proto/teleport/autoupdate/v1/autoupdate.proto:222]: nit:

```suggestion
  // AUTO_UPDATE_AGENT_GROUP_STATE_ROLLEDBACK represents that the group has been rolled back.
```

**Review by codingllama — APPROVED** [review]

**Pull request merged**